src/solvers/newtons_method.jl

```diff
@@ -0,0 +1,76 @@
+export NewtonsMethod
+
+"""
+    NewtonsMethod(; linsolve, convergence_checker, max_iters, update_j)
+
+Solves the equation `f(x) = 0`, given the Jacobian `j(x) = ∂f/∂x`. This is done
+by calling `(::NewtonsMethod)(x, f!, j!)`, where `f!(f, x)` is a function that
+sets `f(x)` in-place and `j!(j, x)` is a function that sets `j(x)` in-place. The
+`x` passed to a `NewtonsMethod` is modified in-place, and its initial value is
+used as a starting guess for the root.
+
+Before using a `NewtonsMethod`, it is necessary to allocate a cache with
+`allocate_cache(::NewtonsMethod, x_prototype, j_prototype)`, where `x_prototype`
+is `similar` to `x` and `f(x)`, and where `j_prototype` is `similar` to `j(x)`.
+Note that `x` and `f(x)` must be `similar` to each other in order for the
+Jacobian to be invertible (i.e., in order for it to be a square matrix).
+"""
+Base.@kwdef struct NewtonsMethod{L, CC <: Union{Nothing, ConvergenceChecker}, C}
+    linsolve::L
+    convergence_checker::CC = nothing
+    max_iters::Int = 1
+    update_j::Bool = false
+    _cache::C = nothing
+end
+
+allocate_cache(newtons_method::NewtonsMethod, x_prototype, j_prototype) =
+    NewtonsMethod(
+        newtons_method.linsolve(Val{:init}, j_prototype, x_prototype),
+        isnothing(newtons_method.convergence_checker) ? nothing :
+            allocate_cache(newtons_method.convergence_checker, x_prototype),
+        newtons_method.max_iters,
+        newtons_method.update_j,
+        (;
+            Δx = similar(x_prototype),
+            f = similar(x_prototype),
+            j = similar(j_prototype),
+        ),
+    )
+
+#=
+We want to solve the equation f(x) = 0, given j(x) := ∂f/∂x.
+Let x[n] denote the value of x on the n-th Newton iteration, and suppose that
+x[n] is converging to some root x′.
+A first-order Taylor series expansion tells us that
+    f(x′) ≈ f(x[n]) + j(x[n]) * (x′ - x[n])
+Since f(x′) = 0, we have that the error on the n-th iteration is roughly
+    x[n] - x′ ≈ Δx[n], where Δx[n] := j(x[n]) \ f(x[n])
+We can then set x[n+1] to be the value of x′ given by this approximation:
+    x[n+1] := x[n] - Δx[n]
+If j changes very slowly, we can approximate j(x[n]) ≈ j(x[0]); i.e, we can use
+the "chord method".
+If a convergence checker is provided, we can use it to determine whether to stop
+iterating on iteration n based on the value x[n] and its error Δx[n]; otherwise,
+we iterate from n = 0 to the maximum number of iterations.
+=#
+function (newtons_method::NewtonsMethod)(x, f!, j!)
```

Comment:
Can we write a `solve` function or something instead of using a functor? Functors can be great, but it can also make it really difficult to find where the solve is actually happening because the variable name could be anything.

Reply:
I was hoping that we could maintain a consistent `allocate_cache` design pattern. In the current implementation, you have some `object` that requires a cache, you call `allocate_cache(object, args...)`, and then you use the object with `object(args...)`. I thought about using a function instead of a functor, but I couldn't come up with a name for that function that will encompass everything that follows this pattern (convergence checker, nonlinear solver, and, in future PRs, the ODE solver and the linear solver). Of course, we could just define a new function for every type of object that follows this pattern (e.g., `run_convergence_checker!` and `run_newtons_method!`), but that might be a bit more confusing for users.